04_Scheduling/04_Locks.md

```diff
@@ -0,0 +1,205 @@
+
+![Tasks execution sequence](Images/taskssequence.png)
+
+Where A, B, and C are the processes being executed, and tasks are prempted everytime they write something to the serial port, what we will have is that the output of the serial will be junk text with A and B outputs overlapping, the picture below describe what can happen in this scenario: 
+
+![Shared Resource Sequence](Images/sharedressequence.png)
+
+This means that a user reading on the serial output (it can be an lcd display, a printer, whatever) he will see something like: 
+
+```
+IWahmi....
+```
+
+While we should expect: 
+
+```
+I am the first string While i am the second
+```
+
+The above scenario happens because both processes (A and B)  have access to the shared resource at the same time, and both of them are writing to it. The junk string above is an example of _race condition_, they are situation in which memory location are accessed concurrently, and at least one of the access is a write. We can imagine that the scenario above can lead to many problems so what we need is a mechanism to protect the shared resource. This is where we introduce the concept of locking.
+
+## Implementing the Lock 
+
+A _lock_ mechanism provide mutual exclusion , ensuring that only a process (or cpu or thread) at a time  can hold the lock and access the shared resource. 
+
+To achieve that we need:
+
+* A lock variable, that is shared betwen  processes/threads who wants to use the shared resource
+* Two new functions to handle the lock called `acquire` and `release` that respectively will acquire the lock to the shared resource blocking other processes to access it, and release it.
+
+For the lock variable a good idea is to create a new datatype, so we can eventually expand it in the future. We will define a data structure with just a single boolean variable for now that will represent the lock.
+
+```c
+
+struct {
+    bool lock;
+} spinlock_t;
+
+typedef struct spinlock_t spinlock_t;
+
+void acquire(spinlock_t *lock);
+void release(spinlock_t *lock);
+```
+
+But how can we ensure mutual exclusion then? The basic idea is pretty straight forward, imagine we update both processes to use the acquire and release function while accessing the the shared resource(be aware in order to have a functionig lock mechanism we should have some kind of memory allocation mechanism implemented). This is the first process:
+
+```c
+ //Process A
+#define SHARED_RESOURCE 0xDEADBEEF 
+
+// Other code doing other stuff
+char string_to_send[] = "I am the first string"
+
+spinlock_t *lock = malloc(sizeof(spinlock_t));
+
+int i = 0;
+acquire(&lock);
+while(i < strlen(string_to_send) {
+    *((int *) shared_resource) = string_to_send[i++];
+}
+release(&lock)
+// Some other code 
+``` 
+
+and below the second one: 
+
+```c
+#define SHARED_RESOURCE 0xDEADBEEF 
+// Other code doing other stuff
+char string_to_send[] = "While i am the second"
+
+spinlock_t *lock = malloc(sizeof(spinlock_t));
```

Comment:
This implementation won't actually protect the shared resource: you have created two *different* locks. So they can both be acquired and allow access, and are functionally independent.

Reply:
A easy solution is to remove this instance of the lock, and then the code that follows will reference the first instance of the lock.

Reply:
ahaha the curse of copying and pasting XD